src/pip/_internal/cli/command_context.py

```diff
@@ -0,0 +1,26 @@
+from contextlib import contextmanager
+
+from pip._vendor.contextlib2 import ExitStack
+
+
+class CommandContextMixIn(object):
+    def __init__(self):
+        super(CommandContextMixIn, self).__init__()
+        self._in_main_context = False
+        self._main_context = ExitStack()
+
+    @contextmanager
+    def main_context(self):
+        assert not self._in_main_context
+
+        self._in_main_context = True
+        try:
+            with self._main_context:
+                yield
+        finally:
+            self._in_main_context = False
+
+    def enter_context(self, context_provider):
```

Comment:
Let's rename this to `add_cleanup_context`, with a docstring explaining how this is meant for registering for cleanups that would happen after exiting from the command, in potentially any order.

Reply:
Later, we can create a `CleanupContextMixin` that would call `.cleanup()` when exiting the context of that object (defining `__enter__` and `__exit__` appropriately). With that change, this method would be able to assert that the things that get passed into it are, in fact, `CleanupContextMixin` subclasses.

Reply:
I think I understand the intent. I would prefer to wait for a short time to see if the use cases we have for this are narrow enough in scope that it is worth creating that abstraction/simplification.

Reply:
That's fair. As long as we're not sticking to this name if there's a different usecase that better fits us here, I'm good.